### PR TITLE
Ignore all ZFS volumes that are part of a Qubes storage pool.

### PR DESCRIPTION
**Pull request by Rudd-O**
Needed for https://github.com/QubesOS/qubes-core-admin/pull/522 to pass tests.

EDIT: I can confirm that the popup about volumes has stopped appearing when a VM starts and stops, whereas without this PR it appears every time a VM starts and stops.

### CODE REVIEW

File: rpm_spec/core-dom0-linux.spec.in [hunk 1]
```diff
@@ -145,11 +146,13 @@ install -D system-config/lvm-cleanup.cron-daily $RPM_BUILD_ROOT/etc/cron.daily/l
 install -d $RPM_BUILD_ROOT%_udevrulesdir
 install -m 644 system-config/00-qubes-ignore-devices.rules $RPM_BUILD_ROOT%_udevrulesdir
 install -m 644 system-config/12-qubes-ignore-lvm-devices.rules $RPM_BUILD_ROOT%_udevrulesdir
+install -m 644 system-config/61-qubes-ignore-zvol-devices.rules $RPM_BUILD_ROOT%_udevrulesdir
 install -m 644 system-config/99z-qubes-mark-ready.rules $RPM_BUILD_ROOT%_udevrulesdir
 install -m 644 -D system-config/disable-lesspipe.sh $RPM_BUILD_ROOT/etc/profile.d/zz-disable-lesspipe.sh
 install -m 755 -D system-config/kernel-grub2.install $RPM_BUILD_ROOT/usr/lib/kernel/install.d/80-grub2.install
 install -m 755 -D system-config/kernel-xen-efi.install $RPM_BUILD_ROOT/usr/lib/kernel/install.d/90-xen-efi.install
 install -m 755 -D system-config/kernel-remove-bls.install $RPM_BUILD_ROOT/usr/lib/kernel/install.d/99-remove-bls.install
+install -m 755 -D system-config/zvol_is_qubes_volume $RPM_BUILD_ROOT%libdir/udev
```
Comment (marmarek) [inline rpm_spec/core-dom0-linux.spec.in]:
```suggestion
install -m 755 -D system-config/zvol_is_qubes_volume $RPM_BUILD_ROOT%libdir/udev/zvol_is_qubes_volume
```

File: rpm_spec/core-dom0-linux.spec.in [hunk 2]
```diff
@@ -53,6 +53,7 @@ Requires:	%{name}-kernel-install
 Requires:	xdotool
 Requires:	createrepo_c
 Requires:	rpm >= 4.14
+Requires:	systemd-udev
```
Comment (Rudd-O) [inline rpm_spec/core-dom0-linux.spec.in:56]: Missing dep.

File: system-config/61-qubes-ignore-zvol-devices.rules [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# Determine if a zvol is a Qubes volume, then ignore it.
+KERNEL=="zd[0-9]*", SUBSYSTEM=="block", ACTION=="add|change", PROGRAM=="/usr/lib/udev/zvol_is_qubes_volume $devnode", ENV{QUBES_IS_PART_OF_QVM_POOL}="%c"
```
Comment (Rudd-O) [inline system-config/61-qubes-ignore-zvol-devices.rules:2]: Program spits out 1 or 0 depending on whether the volume is in fact tagged with the thing.

File: system-config/61-qubes-ignore-zvol-devices.rules [hunk 4]
```diff
@@ -0,0 +1,3 @@
...
+SUBSYSTEM=="block", ACTION=="add|change", ENV{QUBES_IS_PART_OF_QVM_POOL}=="1", ENV{DM_UDEV_DISABLE_DISK_RULES_FLAG}="1", ENV{UDEV_DISABLE_PERSISTENT_STORAGE_RULES_FLAG}="1"
```
Comment (Rudd-O) [inline system-config/61-qubes-ignore-zvol-devices.rules:3]: Is tag `1`?  Then engage protection mode.

File: system-config/zvol_is_qubes_volume [hunk 5]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/bash
+
+# Check for ZFS.
+zvol_id=/usr/lib/udev/zvol_id
+if ! test -x "$zvol_id" ; then zvol_id=/usr/local/lib/udev/zvol_id ; fi
+if ! test -x "$zvol_id" ; then echo 0 ; exit ; fi  # No ZFS.
```
Comment (Rudd-O) [inline system-config/zvol_is_qubes_volume:6]: We continue with the pattern of making ZFS 100% optional.

File: rpm_spec/core-dom0-linux.spec.in (hunk 1 above)
Comment (Rudd-O) [reply rpm_spec/core-dom0-linux.spec.in]: Thank you.  The fix has been pushed.

File: system-config/61-qubes-ignore-zvol-devices.rules (hunk 4 above)
Comment (marmarek) [reply system-config/61-qubes-ignore-zvol-devices.rules:3]: What if the script fails? To be on the safe side, maybe better include `KERNEL=="zd[0-9]*"` here too, but use `ENV{QUBES_IS_PART_OF_QVM_POOL}!="0"` ?

File: system-config/zvol_is_qubes_volume [hunk 6]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/bash
+
```
Comment (marmarek) [inline system-config/zvol_is_qubes_volume:2]: `set -e` ?

Comment (Rudd-O) [reply system-config/zvol_is_qubes_volume:2]: Every line that needed error handling has had it already.  We do not care if zfs list fails, because we evaluate the stdout, not the stderr.

Comment (marmarek) [reply system-config/zvol_is_qubes_volume:2]: The point of `set -e` is about unexpected errors, not expected ones. For example if getting the dataset fails for some reason.

Comment (Rudd-O) [reply system-config/zvol_is_qubes_volume:2]: I've added it.

**Comment by Rudd-O** [comment]
Um, none of the failures seem connected to this PR.  Even the out of memory ones — I can't see any log item at all in the journal that shows the ZFS driver was ever loaded.

**Comment by Rudd-O** [comment]
```
# timestamp 2023-03-06T21:56:08.253234
Traceback (most recent call last):
  File "/usr/lib64/python3.11/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/qubes/tests/integ/basic.py", line 325, in test_202_udev_block_exclude_default
    self.assertVolumesExcludedFromUdev(self.vm)
  File "/usr/lib/python3.11/site-packages/qubes/tests/integ/basic.py", line 273, in assertVolumesExcludedFromUdev
    self._check_udev_for_uuid(private_uuid)
  File "/usr/lib/python3.11/site-packages/qubes/tests/integ/basic.py", line 253, in _check_udev_for_uuid
    self.assertNotIn(uuid_value, udev_file.read(),
AssertionError: '7d60680b-393b-4e2e-858a-8e58f358ffbb' unexpectedly found in 'S:testpool/test-inst-appvm/private\nS:zvol/testpool/test-inst-appvm/private\nS:disk/by-uuid/7d60680b-393b-4e2e-858a-8e58f358ffbb\nS:disk/by-diskseq/115\nI:2184197787\nE:ID_FS_UUID=7d60680b-393b-4e2e-858a-8e58f358ffbb\nE:ID_FS_UUID_ENC=7d60680b-393b-4e2e-858a-8e58f358ffbb\nE:ID_FS_VERSION=1.0\nE:ID_FS_TYPE=ext4\nE:ID_FS_USAGE=filesystem\nE:QUBES_IS_PART_OF_QVM_POOL=0\nG:systemd\nQ:systemd\nV:1\n' : udev parsed filesystem UUID! b230:192
```

Damn.  You're right.  The system has not prescribed that dataset as part of a ZFS Qubes pool.  Will need to dig into this.

**Comment by Rudd-O** [comment]
I need to add logging to `zvol_is_qubes_volume`, because I cannot reproduce that exact situation in my Qubes OS 4.1 system here.

Can you please re-run the OpenQA tests after the incoming commit?  Thanks.

**Comment by Rudd-O** [comment]
This is the fix: https://github.com/QubesOS/qubes-core-admin/pull/522/commits/ffc054a3a24f04fc30e1cce8175339604e385352

It covers the corner case that the tests are clearly exercising — the storage pool is the root pool.

**Comment by Rudd-O** [comment]
```
[2023-03-06T21:37:06.501421Z] [debug] [pid:6074] <<< testapi::assert_script_run(cmd="zpool create -f testpool /dev/sdb1", timeout=90, quiet=undef, fail_message="")
[2023-03-06T21:37:06.501595Z] [debug] [pid:6074] tests/switch_pool.pm:65 called testapi::assert_script_run -> lib/qubesdistribution.pm:306 called testapi::type_string
[2023-03-06T21:37:06.501794Z] [debug] [pid:6074] <<< testapi::type_string(string="zpool create -f testpool /dev/sdb1", max_interval=100, wait_screen_change=0, wait_still_screen=0, timeout=30, similarity_level=47)
[2023-03-06T21:37:11.060323Z] [debug] [pid:6074] tests/switch_pool.pm:65 called testapi::assert_script_run -> lib/qubesdistribution.pm:318 called testapi::type_string
[2023-03-06T21:37:11.060697Z] [debug] [pid:6074] <<< testapi::type_string(string="; echo 1Mmly-\$?- > /dev/hvc0\n", max_interval=100, wait_screen_change=0, wait_still_screen=0, timeout=30, similarity_level=47)
[2023-03-06T21:37:15.219454Z] [debug] [pid:6074] tests/switch_pool.pm:65 called testapi::assert_script_run -> lib/qubesdistribution.pm:320 called testapi::wait_serial
[2023-03-06T21:37:15.219859Z] [debug] [pid:6074] <<< testapi::wait_serial(no_regex=0, record_output=undef, timeout=90, quiet=undef, buffer_size=undef, expect_not_found=0, regexp=qr/1Mmly-\d+-/)
[2023-03-06T21:37:17.322313Z] [debug] [pid:6074] >>> testapi::wait_serial: (?^:1Mmly-\d+-): ok
[2023-03-06T21:37:17.322673Z] [debug] [pid:6074] tests/switch_pool.pm:66 called testapi::assert_script_run
[2023-03-06T21:37:17.322902Z] [debug] [pid:6074] <<< testapi::assert_script_run(cmd="qvm-pool add --option container=testpool pool-test zfs", fail_message="", timeout=90, quiet=undef)
[2023-03-06T21:37:17.323091Z] [debug] [pid:6074] tests/switch_pool.pm:66 called testapi::assert_script_run -> lib/qubesdistribution.pm:306 called testapi::type_string
[2023-03-06T21:37:17.323305Z] [debug] [pid:6074] <<< testapi::type_string(string="qvm-pool add --option container=testpool pool-test zfs", max_interval=100, wait_screen_change=0, wait_still_screen=0, timeout=30, similarity_level=47)
```

Should we as a policy reject the use of root datasets as `container` values for the ZFS pools?

**Comment by Rudd-O** [comment]
@marmarek The revert did not "fail".  We just reverted to the version of the volume snapshotted right after the volume stopped, which *is* the latest version:

* clone and checksum
  * creates snapshot after-import
* start / stop
  * creates snapshot after-stop
* revert
  * reverts to snapshot after-stop

The checksum after revert is going to be the same after stop.

What should we do in this case?

**Comment by Rudd-O** [comment]
@marmarek Some tests are failing with timeouts.  Try increasing the `zfs_arc_max` to 128 MB and see if this speeds up the tests so they don't timeout.  We must find a value that is acceptable, as 64 MB may cause too much slowness.  **Or** we create the pool in a RAM disk, and then it goes at RAM speed, ARC be damned.

**Comment by marmarek** [comment]
Well, if "revert" is no-op most of the time (you can't revert when VM is running), I'd say it's pretty unintuitive behavior. Without explicit revision, it should revert last changes. If you insist on taking snapshot after stop, maybe adjust how revert chooses default revision to revert to?

**Comment by Rudd-O** [comment]
Hm.  What if I changed the behavior of `save_on_stop` volumes to snapshot before start instead of after stop?

**Comment by Rudd-O** [comment]
I believe I've come up with a fix whereby we keep a separate `clean` snapshot which we use to clone from, totally separate from revisions.  Now I need to figure out whether to delete a ZFS snapshot when rolling back to it.

**Review by marmarek — APPROVED** [review]
